azkaban-db/src/main/java/azkaban/db/MySQLDataSource.java

```diff
@@ -58,18 +65,126 @@ public synchronized Connection getConnection() throws SQLException {
          * Every Attempt generates a thread-hanging-time, about 75 seconds, which is hard coded, and can not be changed.
```

Comment:
why does Connection getConnection()  has to be synchronized? Generally I think keeping the synchronize scope as small as possible especially getConnection will potentially hanging for a long period, and given createDataSource().getConnection() is already synchronized, should we not mark this method synchronized?

Reply:
also seems impl. of this method will catch all possible SQLException, when would this method throw SQLException?

Reply:
When the connection can't be established during a failover for example, other threads who are trying to get a connection are likely run into the same issue.  I don't think the overhead of synchronizing this method call is too high.  It's safer to do so when in doubt.

> given createDataSource().getConnection() is already synchronized

Where is this documented?

Reply:
* Did you mean `createDataSource()` is synchronized as well? The benefit of this design can be easily observed in the failover test. When multiple threads are trying to getConnection during DB failover, only one thread keep retrying. It made log very clear. 

* You are likely right. Let me remove the throw Exception signature.

Reply:
createDataSource() is implemented in this PR as synchronized right? getConnection() i might miss it, but we can synchronize on createDataSource().getConnection() only. My point is synchronization on a method is enforcing an object-level lock, in case when getConnection() is hanging, all other synchronized methods(currently there is none, but we might have them in the future ) will be blocked as well.

Reply:
The parent class method signature has `SQLException`. Then we should keep it. I guess `createDataSource().getConnection()` might throw Exception.

Reply:
After thinking, we can remove synchronized and give a try to guarantee normal cases performance.

Reply:
Discussed with @kunkun-tang offline. This change does introduce another network call to check if the DB is readonly each time a connection is requested. Having this method synchronized does have some performance implications. 

org.apache.commons.dbcp2.BasicDataSource's methods seem to handle synchronization already.